src/ngraph/ops/batchnorm.cc

```diff
@@ -0,0 +1,166 @@
+
+  CHECK(channel_axis < batch_data_rank);
+  const size_t channel_axis_length = batch_data_shape[ channel_axis ];
+
+  const ngraph::Shape channel_vector_plus_axes_shape = get_vector_plus_axes_shape(
+      batch_data_rank, channel_axis, channel_axis_length);
+
+  NgraphNodePtr ng_normalized_batch;
+
+  const NgraphNodePtr ng_batch_means =
+    Emitter::ReduceAxes(ng_in_data, {channel_axis}, true, true, ngraph::builder::mean);
+
+  const NgraphNodePtr ng_batch_variances =
+    Emitter::ReduceAxes(ng_in_data, {channel_axis}, true, true,
+        [](const std::shared_ptr<ngraph::Node>& node,
+          const ngraph::AxisSet& axes) {
+        return ngraph::builder::variance(node, axes);
+        });
+
+  const NgraphNodePtr ng_epsilon_shaped = makeConstant(et, channel_vector_plus_axes_shape, epsilon);
+
+  const NgraphNodePtr ng_beta_shaped =
+    ensure_vector_plus_axes_shape(ng_beta, 0, batch_data_rank, channel_axis);
+
+  const NgraphNodePtr ng_gamma_shaped_or_null = ng_maybe_gamma
+    ? ensure_vector_plus_axes_shape(ng_maybe_gamma, 0, batch_data_rank, channel_axis)
+    : NgraphNodePtr{};
+
+  ng_normalized_batch = create_batchnorm_basic_computation_nodes(
+      ng_batch_means, ng_batch_variances, ng_in_data, ng_epsilon_shaped,
+      ng_gamma_shaped_or_null, ng_beta_shaped);
+
+  const NgraphNodePtr ng_batch_means_vector_shaped =
+    ensure_vector_only_shape(ng_batch_means);
+
+  const NgraphNodePtr ng_batch_variances_vector_shaped =
+    ensure_vector_only_shape(ng_batch_variances);
+
+  return std::tuple<NgraphNodePtr, NgraphNodePtr, NgraphNodePtr>{
+    ng_normalized_batch,
+    ng_batch_means_vector_shaped,
+    ng_batch_variances_vector_shaped};
+}
+
+NgraphNodePtr create_batchnorm_inference_without_ngraph_bn_op(
+    const float epsilon,
+    const NgraphNodePtr ng_maybe_gamma,
+    const NgraphNodePtr ng_beta,
+    const NgraphNodePtr ng_in_data,
+    const NgraphNodePtr ng_moving_mean,
+    const NgraphNodePtr ng_moving_var,
+    const size_t channel_axis) {
+  const ngraph::Shape & batch_data_shape = ng_in_data->get_shape();
+  const size_t batch_data_rank = batch_data_shape.size();
+
+  CHECK(channel_axis < batch_data_rank);
+  const size_t channel_axis_length = batch_data_shape[ channel_axis ];
+
+  const NgraphNodePtr ng_mean_shaped =
+    ensure_vector_plus_axes_shape(ng_moving_mean, 0, batch_data_rank, channel_axis);
+
+  const NgraphNodePtr ng_var_shaped =
+    ensure_vector_plus_axes_shape(ng_moving_var, 0, batch_data_rank, channel_axis);
+
+  const ngraph::Shape channel_vector_plus_axes_shape = get_vector_plus_axes_shape(
+      batch_data_rank, channel_axis, channel_axis_length);
+
+  const ngraph::element::Type et = ng_in_data->get_element_type();
+  CHECK(ng_beta->get_element_type() == et);
+
+  const NgraphNodePtr ng_epsilon_shaped = makeConstant(et, channel_vector_plus_axes_shape, epsilon);
+
+  const NgraphNodePtr denom = std::make_shared<ngraph::op::Sqrt>(ng_var_shaped + ng_epsilon_shaped);
```

Comment:
Why don't you use create_batchnorm_basic_computation_nodes here?

Reply:
Nice catch.  Missed it due to editing on a small screen on a bumpy train :)  Will fix.